### ISSUE TITLE
`type:` comments work in other methods but not where defined

### ISSUE THREAD

**Issue by ewerybody** [microsoft/pyright#7537]
## Environment data

-   Language Server version: 2024.3.1 (pyright version 1.1.351, commit 87171fbf)
-   OS and version: Win 10.0.19045 Build 19045 22H2
-   Python version: 3.11.4

## Code Snippet

```python
from PySide6 import QtCore, QtWidgets, QtGui
from somemodule import SomeClass

class SomeSubClass(SomeClass):
    reverted = QtCore.Signal()

    def __init__(self, parent):
        super(SomeSubClass, self).__init__(parent)

        self.view = self.ui.changes_view  # type: QtWidgets.QListView
        self.view.customContextMenuRequested.connect(self.some_function)

    def some_function(self):
        geometry = self.view.geometry()
```

## Repro Steps

1. In `__init__` method of a sub-class definition have instance variable mapped to something Unknown from base class
2. suffix with some dedicated type comment like `# type: QtWidgets.QListView`
3. have another instance method using the instance variable

## Expected behavior

* member variables type is recognised in place
* code/member completion works in next lines
* mouse over popup shows `Unknown | ThatType`
* Go to Definition works on members

## Actual behavior

* member variables type is `Unknown`
* no code/member completion
* mouse over popup only shows `Unknown`
* Go to Definition does not work on members
![image](https://github.com/microsoft/pylance-release/assets/218956/fe17729b-dd0c-4a30-be2a-55d2175d3d20)

BUT:
* It ALL works **in another method**!
  ![image](https://github.com/microsoft/pylance-release/assets/218956/d8838649-16af-40bd-a68c-fcdd0fe989e4)
* It ALL works with Python3 type decoration (like `self.view : QtWidgets.QListView = self.ui.changes_view`) (we're using type comments in this project tho)

So it would just be nice if the type comments work like the Python3 typing method. Voilà
Thank you! 😃
ëRiC

**Comment by ewerybody** [comment]
First, Thank you for the recommendation. But I've already written that type comments are used in this project. I'd rather have it different myself but can't do anything about it atm.

Second, that's not the point. I've tested it with Python 3 style type annotation (like I've also written ^) and everything works like expected 👍: it's marked as `Unknown | DeclaredType` and that's fine! I'd love to write it that way but again I can't.

The problem is that it's just not behaving the same way with the type comment. At least not in the same method!

(That being said: A hackfix would be writing these instance variable definitions in another method and executing it before)

**Comment by erictraut** [comment]
OK, then this has nothing to do with PySide6, unresolved imports, classes that derive from Any, or other methods.

Here's a minimal self-contained code sample that I think demonstrates the problem you're reporting. (The reason we ask for minimal, self-contained code samples is to to eliminate irrelevant aspects so the problem is both clear and reproducible.)

```python
from typing import Any, cast

class Foo:
    def __init__(self):
        self.x = cast(Any, 0)  # type: int
        self.y: int = cast(Any, 0)

        # Note different behaviors here
        reveal_type(self.x)  # Any
        reveal_type(self.y)  # int
```

I agree that the difference in behavior here is unexpected and is likely a bug. I'll investigate further.

Let me know if I interpreted the problem incorrectly.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.358, which I just published.
